### PR TITLE
Use version-script with cmake builds

### LINKED ISSUE #775: ABI changes in 1.0.0  - time for a soname bump?
Describe the bug
----------------
I'm comparing libclamav from 0.103.7 (current Fedora) to 1.0.0-rc2 and it's reporting a number of changes to public symbols.  Below are removed functions marked CLAMAV_PUBLIC.
```
$ abipkgdiff  --d1 clamav-lib-debuginfo-0.103.7-4.fc38.x86_64.rpm --d2 clamav-lib-debuginfo-1.0.0-rc2.x86_64.rpm --devel1 clamav-devel-0.103.7-4.fc38.x86_64.rpm --devel2 clamav-devel-1.0.0-rc2.x86_64.rpm clamav-lib-0.103.7-4.fc38.x86_64.rpm clamav-lib-1.0.0-rc2.x86_64.rpm | grep PUBLIC
    [D] 'function void cl_always_gen_section_hash()'    {cl_always_gen_section_hash@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_countsigs(const char*, unsigned int, unsigned int*)'    {cl_countsigs@@CLAMAV_PUBLIC}
    [D] 'function void cl_cvdfree(cl_cvd*)'    {cl_cvdfree@@CLAMAV_PUBLIC}
    [D] 'function cl_cvd* cl_cvdhead(const char*)'    {cl_cvdhead@@CLAMAV_PUBLIC}
    [D] 'function cl_cvd* cl_cvdparse(const char*)'    {cl_cvdparse@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_cvdverify(const char*)'    {cl_cvdverify@@CLAMAV_PUBLIC}
    [D] 'function void cl_debug()'    {cl_debug@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_addref(cl_engine*)'    {cl_engine_addref@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_compile(cl_engine*)'    {cl_engine_compile@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_free(cl_engine*)'    {cl_engine_free@@CLAMAV_PUBLIC}
    [D] 'function long long int cl_engine_get_num(const cl_engine*, cl_engine_field, int*)'    {cl_engine_get_num@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_engine_get_str(const cl_engine*, cl_engine_field, int*)'    {cl_engine_get_str@@CLAMAV_PUBLIC}
    [D] 'function cl_engine* cl_engine_new()'    {cl_engine_new@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_file_props(cl_engine*, clcb_file_props)'    {cl_engine_set_clcb_file_props@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_hash(cl_engine*, clcb_hash)'    {cl_engine_set_clcb_hash@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_meta(cl_engine*, clcb_meta)'    {cl_engine_set_clcb_meta@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_post_scan(cl_engine*, clcb_post_scan)'    {cl_engine_set_clcb_post_scan@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_pre_cache(cl_engine*, clcb_pre_cache)'    {cl_engine_set_clcb_pre_cache@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_pre_scan(cl_engine*, clcb_pre_scan)'    {cl_engine_set_clcb_pre_scan@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_sigload(cl_engine*, clcb_sigload, void*)'    {cl_engine_set_clcb_sigload@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_add_sample(cl_engine*, clcb_stats_add_sample)'    {cl_engine_set_clcb_stats_add_sample@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_decrement_count(cl_engine*, clcb_stats_decrement_count)'    {cl_engine_set_clcb_stats_decrement_count@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_flush(cl_engine*, clcb_stats_flush)'    {cl_engine_set_clcb_stats_flush@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_get_hostid(cl_engine*, clcb_stats_get_hostid)'    {cl_engine_set_clcb_stats_get_hostid@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_get_num(cl_engine*, clcb_stats_get_num)'    {cl_engine_set_clcb_stats_get_num@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_get_size(cl_engine*, clcb_stats_get_size)'    {cl_engine_set_clcb_stats_get_size@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_remove_sample(cl_engine*, clcb_stats_remove_sample)'    {cl_engine_set_clcb_stats_remove_sample@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_submit(cl_engine*, clcb_stats_submit)'    {cl_engine_set_clcb_stats_submit@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_virus_found(cl_engine*, clcb_virus_found)'    {cl_engine_set_clcb_virus_found@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_set_num(cl_engine*, cl_engine_field, long long int)'    {cl_engine_set_num@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_stats_set_cbdata(cl_engine*, void*)'    {cl_engine_set_stats_set_cbdata@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_set_str(cl_engine*, cl_engine_field, const char*)'    {cl_engine_set_str@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_settings_apply(cl_engine*, const cl_settings*)'    {cl_engine_settings_apply@@CLAMAV_PUBLIC}
    [D] 'function cl_settings* cl_engine_settings_copy(const cl_engine*)'    {cl_engine_settings_copy@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_settings_free(cl_settings*)'    {cl_engine_settings_free@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_stats_enable(cl_engine*)'    {cl_engine_stats_enable@@CLAMAV_PUBLIC}
    [D] 'function int cl_finish_hash(void*, void*)'    {cl_finish_hash@@CLAMAV_PUBLIC}
    [D] 'function void cl_fmap_close(cl_fmap_t*)'    {cl_fmap_close@@CLAMAV_PUBLIC}
    [D] 'function cl_fmap_t* cl_fmap_open_handle(void*, size_t, size_t, clcb_pread, int)'    {cl_fmap_open_handle@@CLAMAV_PUBLIC}
    [D] 'function cl_fmap_t* cl_fmap_open_memory(void*, size_t)'    {cl_fmap_open_memory@@CLAMAV_PUBLIC}
    [D] 'function void cl_hash_destroy(void*)'    {cl_hash_destroy@@CLAMAV_PUBLIC}
    [D] 'function void* cl_hash_init(const char*)'    {cl_hash_init@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_init(unsigned int)'    {cl_init@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_load(const char*, cl_engine*, unsigned int*, unsigned int)'    {cl_load@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_retdbdir()'    {cl_retdbdir@@CLAMAV_PUBLIC}
    [D] 'function unsigned int cl_retflevel()'    {cl_retflevel@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_retver()'    {cl_retver@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scandesc(int, const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*)'    {cl_scandesc@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scandesc_callback(int, const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*, void*)'    {cl_scandesc_callback@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scanfile(const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*)'    {cl_scanfile@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scanfile_callback(const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*, void*)'    {cl_scanfile_callback@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scanmap_callback(cl_fmap_t*, const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*, void*)'    {cl_scanmap_callback@@CLAMAV_PUBLIC}
    [D] 'function void cl_set_clcb_msg(clcb_msg)'    {cl_set_clcb_msg@@CLAMAV_PUBLIC}
    [D] 'function int cl_statchkdir(const cl_stat*)'    {cl_statchkdir@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_statfree(cl_stat*)'    {cl_statfree@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_statinidir(const char*, cl_stat*)'    {cl_statinidir@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_strerror(int)'    {cl_strerror@@CLAMAV_PUBLIC}
    [D] 'function int cl_update_hash(void*, void*, size_t)'    {cl_update_hash@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t lsig_sub_matched(const cli_matcher*, cli_ac_data*, uint32_t, uint32_t, uint32_t, int)'    {lsig_sub_matched@@CLAMAV_PUBLIC}
```
This seems likely to break a number of users of libclamav.

Also, it seems like all of the rust symbols end up in the namespace, but I suppose that there is not much to be done about that at this point.

### PR DESCRIPTION
**Pull request by opoplawski**
This is an attempt to resurrect symbol versioning and fix #775 .  One problem is the new unit tests that link to libclamav and test internal functions do not have access to the symbols.  I started adding them to the symbol map, but I don't think this is the proper solution.  I think instead there perhaps be a static library that is used for the unit tests?

### CODE REVIEW

**Comment by opoplawski** [comment]
We're going to need a check for the existence of version-script I think - there are some examples out there like https://github.com/jrl-umi3218/jrl-cmakemodules/blob/master/version-script.cmake

File: libclamav/CMakeLists.txt [hunk 1]
```diff
@@ -568,6 +568,9 @@ if(ENABLE_SHARED_LIB)
     else()
         target_link_libraries( clamav PUBLIC Iconv::Iconv )
     endif()
+    if(UNIX)
```
Comment (micahsnyder) [inline libclamav/CMakeLists.txt:571]: @opoplawski it appears this breaks the build for macOS with:
```
...
ld: unknown option: --version-script=/Users/runner/work/clamav/clamav/libclamunrar_iface/libclamunrar_iface.map
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libclamunrar_iface/libclamunrar_iface.9.2.0.dylib] Error 1
make[1]: *** [libclamunrar_iface/CMakeFiles/clamunrar_iface.dir/all] Error 2
make: *** [all] Error 2
```

See: https://github.com/Cisco-Talos/clamav/actions/runs/3509775471/jobs/5879108732

Maybe we need to change to 

```
if(UNIX AND NOT APPLE)
```

Same for the other `if(UNIX)`'s in libclamunrar_iface/CMakeLists.txt and libfreshclam/CMakeLists.txt.

**Review by micahsnyder — CHANGES_REQUESTED** [review]

**Comment by micahsnyder** [comment]
I am also seeing failures in our internal jenkins for our statically linked DEB and RPM builds with this failure:
```
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlGzfileOpenW':
xmlIO.c:(.text+0x25a): undefined reference to `gzdopen'
xmlIO.c:(.text+0x298): undefined reference to `gzopen64'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlCreateZMemBuff':
xmlIO.c:(.text+0xa8c): undefined reference to `deflateInit2_'
xmlIO.c:(.text+0xaa3): undefined reference to `crc32'
xmlIO.c:(.text+0xb46): undefined reference to `deflateEnd'
xmlIO.c:(.text+0xbc1): undefined reference to `deflateEnd'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlGzfileClose':
xmlIO.c:(.text+0xbe5): undefined reference to `gzclose'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlGzfileRead':
xmlIO.c:(.text+0xc25): undefined reference to `gzread'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlGzfileWrite':
xmlIO.c:(.text+0xd95): undefined reference to `gzwrite'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlGzfileOpen':
xmlIO.c:(.text+0x16b5): undefined reference to `gzdopen'
xmlIO.c:(.text+0x1714): undefined reference to `gzopen64'
xmlIO.c:(.text+0x1776): undefined reference to `gzdopen'
xmlIO.c:(.text+0x1828): undefined reference to `gzopen64'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `__xmlParserInputBufferCreateFilename':
xmlIO.c:(.text+0x1c93): undefined reference to `gzdirect'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlIOHTTPWrite':
xmlIO.c:(.text+0x2a56): undefined reference to `deflate'
xmlIO.c:(.text+0x2acb): undefined reference to `deflate'
xmlIO.c:(.text+0x2bea): undefined reference to `crc32'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlIOHTTPCloseWrite':
xmlIO.c:(.text+0x32a1): undefined reference to `deflate'
xmlIO.c:(.text+0x33f7): undefined reference to `deflateEnd'
/root/.mussels/install/host-static/lib/libxml2.a(xmlIO.c.o): In function `xmlIOHTTPOpenW':
xmlIO.c:(.text+0x3798): undefined reference to `deflateEnd'
xmlIO.c:(.text+0x37e1): undefined reference to `deflateEnd'
/root/.mussels/install/host-static/lib/libxml2.a(nanohttp.c.o): In function `xmlNanoHTTPFreeCtxt':
nanohttp.c:(.text+0x405): undefined reference to `inflateEnd'
/root/.mussels/install/host-static/lib/libxml2.a(nanohttp.c.o): In function `xmlNanoHTTPRead':
nanohttp.c:(.text+0xe4f): undefined reference to `inflate'
/root/.mussels/install/host-static/lib/libxml2.a(nanohttp.c.o): In function `xmlNanoHTTPMethodRedir':
nanohttp.c:(.text+0x1cfb): undefined reference to `inflateInit2_'
collect2: error: ld returned 1 exit status
[ 99%] Built target clamdtop
unit_tests/CMakeFiles/check_clamav.dir/build.make:402: recipe for target 'unit_tests/check_clamav' failed
make[2]: *** [unit_tests/check_clamav] Error 1
CMakeFiles/Makefile2:2099: recipe for target 'unit_tests/CMakeFiles/check_clamav.dir/all' failed
make[1]: *** [unit_tests/CMakeFiles/check_clamav.dir/all] Error 2
```

I'm not sure why.

**Comment by micahsnyder** [comment]
We have limited time left before the 1.0.0 stable release. We're aiming to publish it one week from today and so are hoping to have everything wrapped up and ready to go by end of business on Wednesday, to account for the american thanksgiving holiday weekend. 
That is -- I have low expectations that we'll be able to solve this in time. 

To resolve your concern about the ABI/namespace change in time for the 1.0.0 release, I have placed a PR to bump the SO version: https://github.com/Cisco-Talos/clamav/pull/778

File: libclamav/CMakeLists.txt (hunk 1 above)
Comment (opoplawski) [reply libclamav/CMakeLists.txt]: Yeah, that's probably a good first-order approximation.  I suspect you'll see issues on other non-GNU platforms though.  I've pushed that change.

**Comment by micahsnyder** [comment]
I just rebased this PR with the upstream `main` branch so it will pass tests in the internal Jenkins pipeline.

**Comment by micahsnyder** [comment]
This PR is also missing using the libclamunrar.map as seen in #816.  I have a commit in a personal/internal branch that i'm testing and if testing goes well, I'll push it here along with pushing a commit that resolves the linking issue I observed with the check_clamav unit test program when linking a static libxml2.a library.

**Review by micahsnyder — APPROVED** [review]

File: libclamav/libclamav.map [hunk 2]
```diff
@@ -61,6 +61,15 @@ CLAMAV_PUBLIC {
...
     lsig_sub_matched;
 };
+CLAMAV_0.104.0 {
+  global:
+    cl_engine_set_clcb_engine_compile_progress;
+    cl_engine_set_clcb_sigload_progress;
+} CLAMAV_PUBLIC;
+CLAMAV_1.0.0 {
+  global:
+    cl_cvdunpack;
```
Comment (micahsnyder) [inline libclamav/libclamav.map:71]: I just realized we should move `cl_engine_set_clcb_file_inspection` here

File: libclamav/libclamav.map [hunk 3]
```diff
@@ -61,6 +61,15 @@ CLAMAV_PUBLIC {
     cl_engine_stats_enable;
     lsig_sub_matched;
 };
+CLAMAV_0.104.0 {
+  global:
+    cl_engine_set_clcb_engine_compile_progress;
+    cl_engine_set_clcb_sigload_progress;
```
Comment (micahsnyder) [inline libclamav/libclamav.map:67]: I just realized we should add `cl_engine_set_clcb_engine_free_progress` here.  It seems to be missing entirely from libclamav.map.

**Review by micahsnyder — CHANGES_REQUESTED** [review]

**Review by micahsnyder — APPROVED** [review]

**Pull request merged**